lxd/storage/drivers/drivers_lvm_volumes.go

```diff
@@ -0,0 +1,732 @@
+		}
+	}
+
+	return nil
+}
+
+// RenameVolume renames a volume and its snapshots.
+func (d *lvm) RenameVolume(vol Volume, newVolName string, op *operations.Operation) error {
+	volDevPath := d.lvmDevPath(d.config["lvm.vg_name"], vol.volType, vol.contentType, vol.name)
+	lvExists, err := d.logicalVolumeExists(volDevPath)
+	if err != nil {
+		return err
+	}
+
+	if !lvExists {
+		return fmt.Errorf("LVM logical volume doesn't exist")
+	}
+
+	return vol.UnmountTask(func(op *operations.Operation) error {
+		snapNames, err := d.VolumeSnapshots(vol, op)
+		if err != nil {
+			return err
+		}
+
+		revert := revert.New()
+		defer revert.Fail()
+
+		// Rename snapshots (change volume prefix to use new parent volume name).
+		for _, snapName := range snapNames {
+			snapVolName := GetSnapshotVolumeName(vol.name, snapName)
+			snapVolDevPath := d.lvmDevPath(d.config["lvm.vg_name"], vol.volType, vol.contentType, snapVolName)
+			newSnapVolName := GetSnapshotVolumeName(newVolName, snapName)
+			newSnapVolDevPath := d.lvmDevPath(d.config["lvm.vg_name"], vol.volType, vol.contentType, newSnapVolName)
+			err = d.renameLogicalVolume(snapVolDevPath, newSnapVolDevPath)
+			if err != nil {
+				return err
+			}
+			revert.Add(func() { d.renameLogicalVolume(newSnapVolDevPath, snapVolDevPath) })
+		}
+
+		// Rename snapshots dir if present.
+		srcSnapshotDir := GetVolumeSnapshotDir(d.name, vol.volType, vol.name)
+		dstSnapshotDir := GetVolumeSnapshotDir(d.name, vol.volType, newVolName)
+		if shared.PathExists(srcSnapshotDir) {
+			err = os.Rename(srcSnapshotDir, dstSnapshotDir)
+			if err != nil {
+				return errors.Wrapf(err, "Error renaming LVM logical volume snapshot directory from '%s' to '%s'", srcSnapshotDir, dstSnapshotDir)
+			}
+			revert.Add(func() { os.Rename(dstSnapshotDir, srcSnapshotDir) })
+		}
+
+		// Rename actual volume.
+		newVolDevPath := d.lvmDevPath(d.config["lvm.vg_name"], vol.volType, vol.contentType, newVolName)
+		err = d.renameLogicalVolume(volDevPath, newVolDevPath)
+		if err != nil {
+			return err
+		}
+		revert.Add(func() { d.renameLogicalVolume(newVolDevPath, volDevPath) })
+
+		// Rename volume dir.
+		srcVolumePath := GetVolumeMountPath(d.name, vol.volType, vol.name)
+		dstVolumePath := GetVolumeMountPath(d.name, vol.volType, newVolName)
+		err = os.Rename(srcVolumePath, dstVolumePath)
+		if err != nil {
+			return errors.Wrapf(err, "Error renaming LVM logical volume mount path from '%s' to '%s'", srcVolumePath, dstVolumePath)
+		}
+		revert.Add(func() { os.Rename(dstVolumePath, srcVolumePath) })
+
+		revert.Success()
+		return nil
+	}, op)
+}
+
+// RestoreVolume restores a volume from a snapshot.
+func (d *lvm) RestoreVolume(vol Volume, snapshotName string, op *operations.Operation) error {
+	// Instantiate snapshot volume from snapshot name.
+	snapVol, err := vol.NewSnapshot(snapshotName)
+	if err != nil {
+		return err
+	}
+
+	// Check snapshot volume exists.
+	if !d.HasVolume(snapVol) {
+		return fmt.Errorf("Snapshot not found")
+	}
```

Comment:
Any reason we need to do the early check?

Reply:
I've removed this, was just belt and braces.